napari/components/dims.py

```diff
@@ -1,265 +1,218 @@
-            current_step=None,
-            axis_labels=None,
-            ndim=None,
-            ndisplay=None,
-            order=None,
-            range=None,
-            last_used=None,
-            deprecated={"axis": "current_step", "camera": "ndisplay"},
+    ndim: int = 2
+    ndisplay: Property[int, None, only_2D_3D] = 2
+    last_used: int = 0
+
+    range: Property[Tuple, None, tuple] = ()
+    current_step: Property[Tuple, None, tuple] = ()
+    order: Property[Tuple, None, tuple] = ()
+    axis_labels: Property[Tuple, None, tuple] = ()
+
+    _scroll_progress: ClassVar[int] = 0
+
+    def __post_init__(self):
+        max_ndim = max(
+            self.ndim,
+            self.ndisplay,
+            len(self.axis_labels),
+            len(self.order),
+            len(self.range),
+            len(self.current_step),
         )
-        self._range = []
-        self._current_step = []
-        self._order = []
-        self._axis_labels = []
-        self._scroll_progress = 0
-        self._last_used = None
-        self._ndisplay = 2 if ndisplay is None else ndisplay
-
-        if ndim is None and order is None and axis_labels is None:
-            ndim = self._ndisplay
-        elif ndim is None and order is None:
-            ndim = len(axis_labels)
-        elif ndim is None and axis_labels is None:
-            ndim = len(order)
-        self.ndim = ndim
-
-        if order is not None:
-            if len(order) != ndim:
-                raise ValueError(
-                    f"Length of order must be identical to ndim."
-                    f" ndim is {ndim} while order is {order}."
-                )
-            self._order = order
-        if axis_labels is not None:
-            if len(axis_labels) != ndim:
-                raise ValueError(
-                    f"Length of axis labels must be identical to ndim."
-                    f" ndim is {ndim} while axis labels is {axis_labels}."
-                )
-            self._axis_labels = list(axis_labels)
+        self._on_ndim_set(max_ndim)
 
-    @property
-    def range(self):
-        """List of 3-tuple: (min, max, step size) of each dimension.
-        """
-        return copy(self._range)
-
-    @property
-    def nsteps(self):
-        """Number of slider steps for each dimension.
-        """
-        return [
-            int((max_val - min_val) // step_size) + 1
-            for min_val, max_val, step_size in self._range
-        ]
-
-    @property
-    def current_step(self):
-        """Tuple of int: value of slider position for each dimension."""
-        return copy(self._current_step)
-
-    @property
-    def point(self):
-        """List of float: value of each dimension."""
-        # The point value is computed from the current_step
-        point = [
-            min_val + step_size * value
-            for (min_val, max_val, step_size), value in zip(
-                self._range, self._current_step
+    def _on_ndim_set(self, ndim):
```

Comment:
I'm probably misunderstanding something and need to read the evented dataclass docs, but how/when does this get called?

Reply:
These get called after the setter, see [here in the code](https://github.com/napari/napari/blob/master/napari/utils/events/dataclass.py#L174-L187) and this comment describing it. https://github.com/napari/napari/pull/1489#discussion_r460423875. I can add a comment